pkg/container/versioned/value.go

```diff
@@ -0,0 +1,375 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package versioned
+
+import (
+	"math"
+	"slices"
+	"sync/atomic"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+type Version uint64
+
+const (
+	// invalidVersion is never found from the value range.
+	// Also used as the upper bound for new values, so that the value is
+	// found when looking with 'maxVersion'
+	invalidVersion = Version(math.MaxUint64)
+
+	// maxVersion in a handle finds all non-removed values
+	maxVersion = Version(math.MaxUint64 - 1)
+
+	maxEmptyHandles = 1024
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "versioned")
+)
+
+type Closer func() bool
+
+// Handle is the object needed for getting values containing the latest version available at the
+// time of getting the Handle.
+type Handle struct {
+	// Handle stores version + 1 so that the zero value translates to invalidVersion
+	versionPlus1 Version
+	// closer releases the handle, can be nil
+	closer Closer
+}
+
+func initHandle(version Version, closer Closer) Handle {
+	return Handle{version + 1, closer}
+}
+
+// AllHandle returns a handle that matches all current/non-removed values
+func AllHandle() Handle {
+	return initHandle(maxVersion, nil)
+}
+
+func (h *Handle) getVersion() Version {
+	return h.versionPlus1 - 1
+}
+
+func (h *Handle) IsValid() bool {
+	return h.versionPlus1 != 0
+}
+
+func (h *Handle) Release() (released bool) {
+	if h.IsValid() {
+		if h.closer != nil {
+			released = h.closer()
+			h.closer = nil
+		}
+		h.versionPlus1 = 0
+	}
+	return released
+}
+
+type atomicVersion struct {
+	version atomic.Uint64
+}
+
+func (a *atomicVersion) Load() Version {
+	return Version(a.version.Load())
+}
+
+func (a *atomicVersion) Store(version Version) {
+	a.version.Store(uint64(version))
+}
+
+func (a *atomicVersion) CompareAndSwap(old, new Version) bool {
+	return a.version.CompareAndSwap(uint64(old), uint64(new))
+}
+
+type versionCount struct {
+	version Version
+	count   int
+}
+
+// Manager manages the versioning for multiple VersionedValues.
+// It is to be stored wherever VersionedValues are stored
+type Manager struct {
+	version atomicVersion
+
+	// cleaner is called with the earliest version that must be kept
+	cleaner func(keepVersion Version)
+
+	// mutex protects 'handles'
+	mutex lock.RWMutex
+
+	// handles is an ordered list of outstanding handle versions with a reference count.
+	// Outdated values can be removed when there are no outstanding handles for it.
+	handles []versionCount
+
+	// latest cleaned version
+	cleanedVersion Version
+}
+
+func NewVersionManager(cleaner func(keepVersion Version)) *Manager {
```

Comment:
I wonder if something like `VersionSet` would be a better name here?

Reply:
I also despise the generic `Manager` as a name, but it is not a container itself, so `Set` would be wrong IMO.

Reply:
Right yeah, `Set` seems wrong indeed. Maybe `Coordinator` could work too, but that's as generic as manager.

Reply:
I just thought that `Coordinator` would be better, came here and saw you suggested it. Reading that days ago must have primed me to this :-)